{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    " ### Libs importing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import time\n",
    "import csv\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Welcome to the CUM zone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def urls_generator(upper_bound):\n",
    "    for i in range(1, upper_bound):\n",
    "        yield str('/home/sirius/BGCproject/mibig_json_2.0/BGC000' + (4-len(str(i)))*'0' + str(i) + '.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def names_definition(n):\n",
    "    counter = 0\n",
    "    names = []\n",
    "    types = []\n",
    "    for url in urls_generator(n):\n",
    "        try:\n",
    "            with open(url) as f:\n",
    "                data = json.load(f)\n",
    "                for elem in data['cluster'].items():\n",
    "                    if 'organism_name' in elem:\n",
    "                        names.append(elem[1])\n",
    "                    if 'biosyn_class' in elem:\n",
    "                        types.append([elem[1]])\n",
    "        except:\n",
    "            counter += 1\n",
    "            f.close()\n",
    "    print(f'{counter} urls didn\\'t exist')\n",
    "    f.close()\n",
    "    for i in range(len(names)):\n",
    "        names[i] = [names[i], types[i]]\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sirius/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (27,28,29,30,31,33) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>relation_to_type_material</th>\n",
       "      <th>asm_not_live_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2014/08/01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>GCA_000001215.4</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.39</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2019/02/28</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>GCA_000001405.28</td>\n",
       "      <td>different</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2020/06/24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>GCA_000001635.9</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2018/03/15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>GCA_000001735.2</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2009/07/15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>GCA_000001905.1</td>\n",
       "      <td>different</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231169</th>\n",
       "      <td>GCF_916700565.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA9543673</td>\n",
       "      <td>CAKAKO000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2866600</td>\n",
       "      <td>2866600</td>\n",
       "      <td>Xenophilus sp. Marseille-Q4582</td>\n",
       "      <td>strain=Marseille-Q4582</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2021/09/30</td>\n",
       "      <td>PRJEB46730</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_916700565.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231170</th>\n",
       "      <td>GCF_916856725.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA9985269</td>\n",
       "      <td>CAKASJ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2866580</td>\n",
       "      <td>2866580</td>\n",
       "      <td>Haemophilus sp. Marseille-Q0026</td>\n",
       "      <td>strain=Marseille-Q0026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2021/10/08</td>\n",
       "      <td>PRJEB46705</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_916856725.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231171</th>\n",
       "      <td>GCF_916858875.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA9962206</td>\n",
       "      <td>CAKASN000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2866594</td>\n",
       "      <td>2866594</td>\n",
       "      <td>Staphylococcus sp. Marseille-Q1834</td>\n",
       "      <td>strain=Marseille-Q1834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2021/10/08</td>\n",
       "      <td>PRJEB46723</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_916858875.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231172</th>\n",
       "      <td>GCF_916862695.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA9985268</td>\n",
       "      <td>CAKASP000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2866579</td>\n",
       "      <td>2866579</td>\n",
       "      <td>Haemophilus sp. Marseille-P9675</td>\n",
       "      <td>strain=Marseille-P9675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2021/10/08</td>\n",
       "      <td>PRJEB46706</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_916862695.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231173</th>\n",
       "      <td>GCF_918593995.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA10558844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>40324</td>\n",
       "      <td>40324</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stenotrophomonas maltophilia 1800</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2021/11/01</td>\n",
       "      <td>ASM91859399v1</td>\n",
       "      <td>Genetique Moleculaire</td>\n",
       "      <td>Genomique</td>\n",
       "      <td>Microbiologie (GMGM)</td>\n",
       "      <td>Universite de Strasbourg</td>\n",
       "      <td>GCA_918593995.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231174 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       # assembly_accession   bioproject      biosample         wgs_master  \\\n",
       "0           GCF_000001215.4     PRJNA164   SAMN02803731                NaN   \n",
       "1          GCF_000001405.39     PRJNA168            NaN                NaN   \n",
       "2          GCF_000001635.27     PRJNA169            NaN                NaN   \n",
       "3           GCF_000001735.4     PRJNA116   SAMN03081427                NaN   \n",
       "4           GCF_000001905.1   PRJNA70973   SAMN02953622     AAGU00000000.3   \n",
       "...                     ...          ...            ...                ...   \n",
       "231169      GCF_916700565.1  PRJNA224116   SAMEA9543673  CAKAKO000000000.1   \n",
       "231170      GCF_916856725.1  PRJNA224116   SAMEA9985269  CAKASJ000000000.1   \n",
       "231171      GCF_916858875.1  PRJNA224116   SAMEA9962206  CAKASN000000000.1   \n",
       "231172      GCF_916862695.1  PRJNA224116   SAMEA9985268  CAKASP000000000.1   \n",
       "231173      GCF_918593995.1  PRJNA224116  SAMEA10558844                NaN   \n",
       "\n",
       "              refseq_category    taxid  species_taxid  \\\n",
       "0            reference genome     7227           7227   \n",
       "1            reference genome     9606           9606   \n",
       "2            reference genome    10090          10090   \n",
       "3            reference genome     3702           3702   \n",
       "4       representative genome     9785           9785   \n",
       "...                       ...      ...            ...   \n",
       "231169                     na  2866600        2866600   \n",
       "231170                     na  2866580        2866580   \n",
       "231171                     na  2866594        2866594   \n",
       "231172                     na  2866579        2866579   \n",
       "231173                     na    40324          40324   \n",
       "\n",
       "                             organism_name      infraspecific_name  \\\n",
       "0                  Drosophila melanogaster                     NaN   \n",
       "1                             Homo sapiens                     NaN   \n",
       "2                             Mus musculus                     NaN   \n",
       "3                     Arabidopsis thaliana        ecotype=Columbia   \n",
       "4                       Loxodonta africana                     NaN   \n",
       "...                                    ...                     ...   \n",
       "231169      Xenophilus sp. Marseille-Q4582  strain=Marseille-Q4582   \n",
       "231170     Haemophilus sp. Marseille-Q0026  strain=Marseille-Q0026   \n",
       "231171  Staphylococcus sp. Marseille-Q1834  strain=Marseille-Q1834   \n",
       "231172     Haemophilus sp. Marseille-P9675  strain=Marseille-P9675   \n",
       "231173        Stenotrophomonas maltophilia                     NaN   \n",
       "\n",
       "                                  isolate  ... genome_rep seq_rel_date  \\\n",
       "0                                     NaN  ...       Full   2014/08/01   \n",
       "1                                     NaN  ...       Full   2019/02/28   \n",
       "2                                     NaN  ...       Full   2020/06/24   \n",
       "3                                     NaN  ...       Full   2018/03/15   \n",
       "4                              ISIS603380  ...       Full   2009/07/15   \n",
       "...                                   ...  ...        ...          ...   \n",
       "231169                                NaN  ...       Full   2021/09/30   \n",
       "231170                                NaN  ...       Full   2021/10/08   \n",
       "231171                                NaN  ...       Full   2021/10/08   \n",
       "231172                                NaN  ...       Full   2021/10/08   \n",
       "231173  Stenotrophomonas maltophilia 1800  ...       Full   2021/11/01   \n",
       "\n",
       "                      asm_name  \\\n",
       "0       Release 6 plus ISO1 MT   \n",
       "1                   GRCh38.p13   \n",
       "2                       GRCm39   \n",
       "3                     TAIR10.1   \n",
       "4                    Loxafr3.0   \n",
       "...                        ...   \n",
       "231169              PRJEB46730   \n",
       "231170              PRJEB46705   \n",
       "231171              PRJEB46723   \n",
       "231172              PRJEB46706   \n",
       "231173           ASM91859399v1   \n",
       "\n",
       "                                                submitter   gbrs_paired_asm  \\\n",
       "0       The FlyBase Consortium/Berkeley Drosophila Gen...   GCA_000001215.4   \n",
       "1                             Genome Reference Consortium  GCA_000001405.28   \n",
       "2                             Genome Reference Consortium   GCA_000001635.9   \n",
       "3             The Arabidopsis Information Resource (TAIR)   GCA_000001735.2   \n",
       "4                                         Broad Institute   GCA_000001905.1   \n",
       "...                                                   ...               ...   \n",
       "231169                       IHU - Mediterranee Infection   GCA_916700565.1   \n",
       "231170                       IHU - Mediterranee Infection   GCA_916856725.1   \n",
       "231171                       IHU - Mediterranee Infection   GCA_916858875.1   \n",
       "231172                       IHU - Mediterranee Infection   GCA_916862695.1   \n",
       "231173                              Genetique Moleculaire         Genomique   \n",
       "\n",
       "              paired_asm_comp  \\\n",
       "0                   identical   \n",
       "1                   different   \n",
       "2                   identical   \n",
       "3                   identical   \n",
       "4                   different   \n",
       "...                       ...   \n",
       "231169              identical   \n",
       "231170              identical   \n",
       "231171              identical   \n",
       "231172              identical   \n",
       "231173   Microbiologie (GMGM)   \n",
       "\n",
       "                                                 ftp_path  \\\n",
       "0       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "...                                                   ...   \n",
       "231169  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "231170  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "231171  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "231172  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "231173                           Universite de Strasbourg   \n",
       "\n",
       "       excluded_from_refseq relation_to_type_material  \\\n",
       "0                       NaN                       NaN   \n",
       "1                       NaN                       NaN   \n",
       "2                       NaN                       NaN   \n",
       "3                       NaN                       NaN   \n",
       "4                       NaN                       NaN   \n",
       "...                     ...                       ...   \n",
       "231169                  NaN                       NaN   \n",
       "231170                  NaN                       NaN   \n",
       "231171                  NaN                       NaN   \n",
       "231172                  NaN                       NaN   \n",
       "231173      GCA_918593995.1                 identical   \n",
       "\n",
       "                                        asm_not_live_date  \n",
       "0                                                      na  \n",
       "1                                                      na  \n",
       "2                                                      na  \n",
       "3                                                      na  \n",
       "4                                                      na  \n",
       "...                                                   ...  \n",
       "231169                                                 na  \n",
       "231170                                                 na  \n",
       "231171                                                 na  \n",
       "231172                                                 na  \n",
       "231173  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "\n",
       "[231174 rows x 23 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/home/sirius/BGCproject/assembly_summary_refseq.csv') as fuck:\n",
    "    data = pd.read_csv(fuck)\n",
    "    fuckingdf = pd.DataFrame(data)\n",
    "    fuck.close()\n",
    "fuckingdf = fuckingdf[fuckingdf.columns.drop(list(fuckingdf.filter(regex='Unnamed')))]\n",
    "fuckingdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(231174, 23)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fdf = np.array(fuckingdf)\n",
    "fdf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['  Monterrey', ' 2200 Copenhagen N',\n",
       "       \" 302 Katsura Int'tech Center Kyotodaigaku-Katsura\", ...,\n",
       "       'identical', 'na', 'nan'], dtype='<U131')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = []\n",
    "for line in fdf:\n",
    "    l.append(np.unique(line[19]))\n",
    "np.unique(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([' 1 lactose operon', ' Braunschweig', ' Edinburgh', '13', '141',\n",
       "       '230', '231', '27:l', '5304', '7', 'CECT 8001', 'Chromosome',\n",
       "       'Complete Genome', 'Contig', 'DSM 25260', 'KCTC 23692', 'Scaffold',\n",
       "       'latest', 'n', 'nan', 'strain=BCW_1534', 'strain=BCW_1538',\n",
       "       'strain=CFSAN044923', 'strain=CFSAN044974', 'strain=CFSAN044982',\n",
       "       'strain=CFSAN044983', 'strain=CFSAN044987', 'strain=CFSAN045004',\n",
       "       'strain=CFSAN045006', 'strain=CFSAN045015', 'strain=CFSAN045030',\n",
       "       'strain=CFSAN045036', 'strain=CFSAN045038', 'strain=CFSAN051017',\n",
       "       'strain=CFSAN051043', 'strain=HAZEL-066', 'strain=HAZEL-067',\n",
       "       'strain=HAZEL-070', 'strain=HAZEL-071', 'strain=HAZEL-072',\n",
       "       'strain=HAZEL-202', 'strain=HAZEL-203', 'strain=PNCS000211',\n",
       "       'strain=PNCS007087', 'strain=PNCS007098', 'strain=PNCS014846',\n",
       "       'strain=PNCS014847', 'strain=PNCS014848', 'strain=PNCS014849',\n",
       "       'strain=PNCS014850', 'strain=PNCS014852', 'strain=PNCS014853',\n",
       "       'strain=PNCS014855', 'strain=PNCS014857', 'strain=PNCS014858',\n",
       "       'strain=PNCS014859', 'strain=PNCS014860', 'strain=PNCS014861',\n",
       "       'strain=PNCS014863', 'strain=PNCS014864', 'strain=PNCS014865',\n",
       "       'strain=PNCS014866', 'strain=PNCS014867', 'strain=PNCS014868',\n",
       "       'strain=PNCS014873', 'strain=PNCS014875', 'strain=PNCS014876',\n",
       "       'strain=PNCS014880', 'strain=PNCS014881', 'strain=SA20070548',\n",
       "       'strain=SA20082869', 'strain=SA20143792', 'x'], dtype='<U32')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = []\n",
    "for line in fdf:\n",
    "    q.append(np.unique(line[11]))\n",
    "np.unique(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ncbi_parsing(name):\n",
    "    complete_genomes, chromosomes = [], []\n",
    "    scaffolds, contigs = [], []\n",
    "    for line in fdf:\n",
    "        try:     # 7 -- organism name, 11 -- assembly level\n",
    "            if name in line[7]:\n",
    "                if re.search('Complete Genome', line[11]) and re.search('ftp', line[19]):\n",
    "                    complete_genomes.append([name, 'Complete Genome', line[19] + '/' + line[19].split('/')[-1] + '_genomic.fna.gz']) #link in line[19]\n",
    "                elif re.search('Chromosome', line[11]) and re.search('ftp', line[19]): #'Chromosome' in line[11]:\n",
    "                    chromosomes.append([name, 'Chromosome', line[19] + '/' + line[19].split('/')[-1] + '_genomic.fna.gz'])\n",
    "                elif re.search('Scaffold', line[11]) and re.search('ftp', line[19]): #'Scaffold' in line[11]:\n",
    "                    scaffolds.append([name, 'Scaffold', line[19] + '/' + line[19].split('/')[-1] + '_genomic.fna.gz'])\n",
    "                elif re.search('Contig', line[11]) and re.search('ftp', line[19]):\n",
    "                    contigs.append([name, 'Contig', line[19] + '/' + line[19].split('/')[-1] + '_genomic.fna.gz'])\n",
    "        except:\n",
    "            pass\n",
    "            #print(f'Problem with a {name} occured')\n",
    "    if len(complete_genomes) != 0: \n",
    "        return complete_genomes\n",
    "    elif len(chromosomes) != 0: \n",
    "        return chromosomes\n",
    "    elif len(scaffolds) != 0: \n",
    "        return scaffolds\n",
    "    elif len(contigs) != 0:\n",
    "        return contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129 urls didn't exist\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-da0a6558e702>:16: DeprecationWarning: NotImplemented should not be used in a boolean context\n",
      "  for elem in filter(None. __ne__, array):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "853\n",
      "Time spent: 38.57214593887329\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "official_names = names_definition(n=2040)\n",
    "clean_names = []\n",
    "for data in official_names:\n",
    "    if data not in clean_names:\n",
    "        clean_names.append(data)\n",
    "    else:\n",
    "        for line in clean_names:\n",
    "            if data in line:\n",
    "                clean_names[1].append(data[1])\n",
    "\n",
    "array = []\n",
    "for appelation in clean_names:\n",
    "    array.append(ncbi_parsing(name=appelation[0]))\n",
    "clean_array = []\n",
    "for elem in filter(None. __ne__, array):\n",
    "    clean_array.append(elem)\n",
    "for name in clean_names:\n",
    "    for element in clean_array:\n",
    "        if re.search(name[0], element[0][0]) and element[0][2] not in name:\n",
    "            name.append(element[0][1])\n",
    "            name.append(element[0][2])\n",
    "\n",
    "final_table = []\n",
    "for element in clean_names:\n",
    "    if len(element) > 2:\n",
    "        final_table.append(element)\n",
    "\n",
    "for line in final_table:\n",
    "    for data in official_names:\n",
    "        if data[0].strip() == line[0].strip():\n",
    "        #if re.search(data[0], line[0]):\n",
    "            line[1].append(data[1].copy())\n",
    "    line[1].pop(0)\n",
    "    \n",
    "print(len(final_table))\n",
    "print(f'Time spent: {time.time() - start_time}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Alternaria alternata',\n",
       "  [[['Polyketide']],\n",
       "   [['NRP']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['NRP']],\n",
       "   [['Polyketide']]],\n",
       "  'Complete Genome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/874/585/GCF_000874585.1_ViralMultiSegProj30367/GCF_000874585.1_ViralMultiSegProj30367_genomic.fna.gz'],\n",
       " ['Aspergillus flavus',\n",
       "  [[['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['NRP', 'Polyketide']]],\n",
       "  'Chromosome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/014/117/465/GCF_014117465.1_ASM1411746v1/GCF_014117465.1_ASM1411746v1_genomic.fna.gz'],\n",
       " ['Aspergillus ochraceoroseus',\n",
       "  [[['Polyketide']]],\n",
       "  'Contig',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/846/915/GCF_002846915.1_Aspergillus_ochraceoroseus_IBT_24754_v1.1/GCF_002846915.1_Aspergillus_ochraceoroseus_IBT_24754_v1.1_genomic.fna.gz'],\n",
       " ['Aspergillus nidulans FGSC A4',\n",
       "  [[['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['NRP']],\n",
       "   [['Terpene']],\n",
       "   [['Terpene']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP']],\n",
       "   [['NRP']],\n",
       "   [['NRP']],\n",
       "   [['NRP']],\n",
       "   [['NRP']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']]],\n",
       "  'Scaffold',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/149/205/GCF_000149205.2_ASM14920v2/GCF_000149205.2_ASM14920v2_genomic.fna.gz'],\n",
       " ['Sorangium cellulosum',\n",
       "  [[['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']]],\n",
       "  'Complete Genome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/067/165/GCF_000067165.1_ASM6716v1/GCF_000067165.1_ASM6716v1_genomic.fna.gz',\n",
       "  'Complete Genome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/418/325/GCF_000418325.1_ASM41832v1/GCF_000418325.1_ASM41832v1_genomic.fna.gz'],\n",
       " ['Streptomyces nodosus',\n",
       "  [[['Other']], [['RiPP']]],\n",
       "  'Complete Genome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/008/704/995/GCF_008704995.1_ASM870499v1/GCF_008704995.1_ASM870499v1_genomic.fna.gz'],\n",
       " ['Streptomyces eurythermus',\n",
       "  [[['Polyketide']], [['Polyketide']]],\n",
       "  'Scaffold',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/014/649/115/GCF_014649115.1_ASM1464911v1/GCF_014649115.1_ASM1464911v1_genomic.fna.gz'],\n",
       " ['Actinosynnema pretiosum subsp. auranticum',\n",
       "  [[['Polyketide']]],\n",
       "  'Contig',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/013/387/285/GCF_013387285.1_ASM1338728v1/GCF_013387285.1_ASM1338728v1_genomic.fna.gz'],\n",
       " ['Stigmatella aurantiaca DW4/3-1',\n",
       "  [[['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']],\n",
       "   [['NRP', 'Polyketide']]],\n",
       "  'Complete Genome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/165/485/GCF_000165485.1_ASM16548v1/GCF_000165485.1_ASM16548v1_genomic.fna.gz'],\n",
       " ['Streptomyces avermitilis',\n",
       "  [[['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Polyketide']],\n",
       "   [['Terpene']],\n",
       "   [['Other']],\n",
       "   [['Other']],\n",
       "   [['Other']],\n",
       "   [['NRP', 'Polyketide']]],\n",
       "  'Complete Genome',\n",
       "  'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/009/765/GCF_000009765.2_ASM976v2/GCF_000009765.2_ASM976v2_genomic.fna.gz']]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_table[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for line in final_table:\n",
    "#     for data in official_names:\n",
    "#         if re.search(data[0], line[0]):\n",
    "#             line[1].append(data[1].copy())\n",
    "#     line[1].pop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.DataFrame(final_table)\n",
    "deselectlist =['organism_name', 'biosyn_class', 'ftp_path']\n",
    "selectlist =[x for x in data.columns if x not in deselectlist]\n",
    "datatowrite = data[selectlist]\n",
    "\n",
    "datatowrite.to_csv('new.csv', index=False)\n",
    "\n",
    "# final = pd.DataFrame(final_table)\n",
    "# final.to_csv('check.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
